Add database code for conversation CRUD

igorveselinovic opened the pull request:
Resolves #5 

Adding the database logic for the API handlers for the CRUD of conversations.

One thing to note is that I used transactions for the creation of a conversation and the deletion of a conversation. What I ended up doing to get this to work was to make the `CreateConversation` and `DeleteConversation` methods include the code for creating/deleting the conversation row and also the associated user mapping(s). There's a `CreateUserConversationMapping` method also defined, but nothing calls it right now. It will eventually be used when writing the code for adding users to a conversation. So, there is some weird sorta-code-duplication that's not ideal that might be worth discussing.

igorveselinovic committed:
Add Datastore interface to wrap db connection

Using MySQL database driver

igorveselinovic committed:
Add struct for passing db connection to handlers

The Env struct contains the db connection (and will contain any future
application-level items) which is needed by the HTTP request handlers.
The handlers are defined as methods against the Env struct.

igorveselinovic committed:
Add DB logic for CRUD of conversations

This mainly deals with the conversations table but also does some stuff
with the users_to_conversations table:
* Creation of a conversation automatically adds the current user as an
OWNER
* Reading and updating of a conversation checks that the current user is
in the conversation
* Deletion of a conversation checks that the current user is the OWNER
of the conversation

A lot of repeated code at the moment that can be put into helper
functions.

igorveselinovic committed:
Add helpers for handlers

These helpers are for verifying that a conversation exists and verifying
that a user is in a conversation.

igorveselinovic committed:
Fix inconsistent variable naming

reqBody -> reqConversation

igorveselinovic committed:
Add field checking to POST handler

Also updated some error log messages

igorveselinovic committed:
Add logging for database queries

igorveselinovic committed:
Change function names to be more accurate

checkConversation -> getConversation
checkMapping -> getMapping

igorveselinovic committed:
Update handler function names to be more accurate

New names reflect the fact that these endpoints operate on a single
conversation.

igorveselinovic committed:
Add some comments

Also renamed a file to have a more accurate name

igorveselinovic committed:
Modify some database code

* creating a mapping doesn't return an int64 anymore
* updating a conversation expects a full conversation struct with no
  missing fields
* creating and deleting conversations and the associated user(s) are now
  done in transactions to ensure atomicity (might want to restructure
  this code later to be a bit nicer)

igorveselinovic committed:
Add doc for conversation merge method

igorveselinovic committed:
Add a check for integer IDs in the URL path

Thao-Tran commented on handlers/handlers.go:
```diff
@@ -29,60 +37,220 @@ func parseReqBody(w http.ResponseWriter, body io.ReadCloser, bodyObj *models.Con
+	// TODO: Create HTML file
+
+	reqConversation.ID = conversationID
+	location := fmt.Sprintf("/api/conversations/%d", conversationID)
```
The location might have to be `/ether/api/conversations/{conversation_
I'd}` depending on how we plan on implementing the return in `heimdall`. Right now `heimdall` is just blindly passing responses on. We could make it so it tacks on the service name in front of the `Location` header if it is set but it might not be worth the extra effort.

igorveselinovic replied:
If we make the `Location` header `/ether/api/conversations/{conversation_id}`, then I think that the routes should also start with `/ether/api/...`, so that there isn't a discrepancy between the two.

Thao-Tran replied:
Hmmm that's true. We can't make the routes have the service name though 'cause then the routing from `heimdall` would get weird, ex `/ether/ether/api/...`. Can you make an issue for `heimdall` then to add the service name to the beginning of the `Location` header of a response if it exists?

igorveselinovic replied:
Does `heimdall` need to remove the service name from the front of any route it receives? Like if it receives a request to `/ether/api/conversations`, can it not just forward to that same path but at `ether`'s location?

Thao-Tran replied:
Oh it doesn't remove the service name from the route though, it just uses the service name as like the DNS nickname thing so if it sends a request to `ether/api/...`, `ether` would receive it as `/api/...`. So if the route was `/ether/api/...` then the client request URL would either have to look like `/ether/ether/api/...` or it would be `/ether/api...` and `heimdall` would have to add `ether` to the start of the request.

Thao-Tran commented on handlers/handlers.go:
```diff
@@ -29,60 +37,220 @@ func parseReqBody(w http.ResponseWriter, body io.ReadCloser, bodyObj *models.Con
-	if err := parseReqBody(w, r.Body, reqBody); err != nil {
+	if reqConversation.Name == "" || reqConversation.Description == nil {
+		errMsg := "Request body is missing field(s)"
+		fmt.Println(errMsg)
```
```suggestion	
		log.Println(errMsg)
```

Thao-Tran commented on handlers/handlers.go:
```diff
@@ -29,60 +37,220 @@ func parseReqBody(w http.ResponseWriter, body io.ReadCloser, bodyObj *models.Con
 
-// PostConversationsHandler creates a new conversation
-func PostConversationsHandler(w http.ResponseWriter, r *http.Request) {
+func getConversation(w http.ResponseWriter, db models.Datastore, id int64) (*models.Conversation, error) {
```
Might be good to make this a method of the `Env` struct since it depends on the same `models.Datastore` instance anyway.

igorveselinovic replied:
My reasoning for not doing this was that since it does HTTP-specific stuff, it didn't really make sense to be part of the `Datastore` interface.

Thao-Tran replied:
Oh I didn't mean to make it a part of the `Datastore` interface, but to make it a method of the `Env` struct.

igorveselinovic replied:
Sorry, I misunderstood. Yeah, I'll make that change.

Thao-Tran commented on handlers/handlers.go:
```diff
@@ -29,60 +37,220 @@ func parseReqBody(w http.ResponseWriter, body io.ReadCloser, bodyObj *models.Con
+	return conversation, nil
+}
+
+func getMapping(w http.ResponseWriter, db models.Datastore, userID, convID int64) (*models.UserConversationMapping, error) {
```
Same as my comment on `getConversation`.

Thao-Tran commented on handlers/handlers.go:
```diff
@@ -29,60 +37,220 @@ func parseReqBody(w http.ResponseWriter, body io.ReadCloser, bodyObj *models.Con
-	reqBody := &models.Conversation{}
-	if err := parseReqBody(w, r.Body, reqBody); err != nil {
+	if reqConversation.Name == "" && reqConversation.Description == nil {
+		errMsg := "Request body has neither \"name\" nor \"description\""
```
You can use backticks instead of double quotes so you don't have to escape the double quotes inside the message.

igorveselinovic replied:
I'll change all these.

Thao-Tran commented on handlers/handlers.go:
```diff
@@ -29,60 +37,220 @@ func parseReqBody(w http.ResponseWriter, body io.ReadCloser, bodyObj *models.Con
-	if err := parseReqBody(w, r.Body, reqBody); err != nil {
+	if reqConversation.Name == "" && reqConversation.Description == nil {
+		errMsg := "Request body has neither \"name\" nor \"description\""
+		fmt.Println(errMsg)
```
```suggestion
		log.Println(errMsg)
```

Thao-Tran commented on models/conversation.go:
```diff
@@ -0,0 +1,171 @@
+		return -1, err
+	}
+	if rowCount, err := res.RowsAffected(); err == nil {
+		log.Printf("Created %d row(s) in \"%s\"", rowCount, conversationsTable)
```
Same thing with the backticks in the earlier comment.

Thao-Tran commented on models/conversation.go:
```diff
@@ -0,0 +1,171 @@
+		}
+		return nil, err
+	}
+	log.Printf("Read 1 row from \"%s\"", conversationsTable)
```
Backticks thing.

Thao-Tran commented on models/conversation.go:
```diff
@@ -0,0 +1,171 @@
+	res, err := db.Exec(b.String())
+	if err == nil {
+		if rowCount, err := res.RowsAffected(); err == nil {
+			log.Printf("Updated %d row(s) in \"%s\"", rowCount, conversationsTable)
```
Backticks sitch.

Thao-Tran commented on models/conversation.go:
```diff
@@ -0,0 +1,171 @@
+	res, err = tx.Exec(queryString)
+	if err == nil {
+		if rowCount, err := res.RowsAffected(); err == nil {
+			log.Printf("Deleted %d row(s) from \"%s\"", rowCount, conversationsTable)
```
Backticks and whatnot.

Thao-Tran commented on models/conversation.go:
```diff
@@ -0,0 +1,171 @@
+		return -1, err
+	}
+	if rowCount, err := res.RowsAffected(); err == nil {
+		log.Printf("Created %d row(s) in \"%s\"", rowCount, mappingsTable)
```
Backticks.

Thao-Tran commented on models/user_conversation_mapping.go:
```diff
@@ -0,0 +1,97 @@
+	if err != nil {
+		return err
+	}
+	if rowCount, err := res.RowsAffected(); err == nil {
```
Does this error just not matter enough to be handled or logged?

igorveselinovic replied:
So the thing with this error is that even if `res.RowsAffected()` returns an error, the `INSERT` still succeeded, so I didn't want to make the function return an error, when the stuff that actually matters already succeeded. That being said, the error should at least be logged, yeah.

Thao-Tran commented on models/user_conversation_mapping.go:
```diff
@@ -0,0 +1,97 @@
+		return err
+	}
+	if rowCount, err := res.RowsAffected(); err == nil {
+		log.Printf("Created %d row(s) in \"%s\"", rowCount, mappingsTable)
```
Backticks.

Thao-Tran reviewed (changes requested):
Also, I didn't want to write this for all of your queries, but what are your thoughts on preparing a `Stmt` for each query when the database connection is created and reusing them in the DB methods? Apparently they're safe for concurrent use with goroutines and it'll save a step since how `QueryRow` and `Exec` work is by making a `Stmt` using the query string and executing it.

igorveselinovic committed:
Fix print to use log

Co-Authored-By: Thao-Tran Le-Phuong <Thao-Tran@users.noreply.github.com>

igorveselinovic committed:
Fix print to use log

Co-Authored-By: Thao-Tran Le-Phuong <Thao-Tran@users.noreply.github.com>

igorveselinovic commented:
> Also, I didn't want to write this for all of your queries, but what are your thoughts on preparing a `Stmt` for each query when the database connection is created and reusing them in the DB methods? Apparently they're safe for concurrent use with goroutines and it'll save a step since how `QueryRow` and `Exec` work is by making a `Stmt` using the query string and executing it.

My understanding of how `db.Exec()` and `db.QueryRow()` work is that they only create a prepared statement if you do something like `db.Exec("INSERT INTO test.hello(world) VALUES(?)", "hello world!")` (i.e, using placeholders), as opposed to the code in this PR which is just passing the pre-built string, so it shouldn't be creating any prepared statements. I think we can also set [this parameter](https://github.com/go-sql-driver/mysql#interpolateparams) to `true` to explicitly prevent the extra round-trips.

Still though, prepared statements could improve performance, and also the way I'm building the queries right now doesn't protect against SQL injection attacks, so that might be the best way to protect against that.

Maybe for now we should just enable the `interpolateParams`, and use the `?` placeholders, which should protect against SQL injection attacks, and then make an issue for switching over to using prepared statements?

Thao-Tran commented:
> > Also, I didn't want to write this for all of your queries, but what are your thoughts on preparing a `Stmt` for each query when the database connection is created and reusing them in the DB methods? Apparently they're safe for concurrent use with goroutines and it'll save a step since how `QueryRow` and `Exec` work is by making a `Stmt` using the query string and executing it.
> 
> My understanding of how `db.Exec()` and `db.QueryRow()` work is that they only create a prepared statement if you do something like `db.Exec("INSERT INTO test.hello(world) VALUES(?)", "hello world!")` (i.e, using placeholders), as opposed to the code in this PR which is just passing the pre-built string, so it shouldn't be creating any prepared statements. I think we can also set [this parameter](https://github.com/go-sql-driver/mysql#interpolateparams) to `true` to explicitly prevent the extra round-trips.
> 
> Still though, prepared statements could improve performance, and also the way I'm building the queries right now doesn't protect against SQL injection attacks, so that might be the best way to protect against that.
> 
> Maybe for now we should just enable the `interpolateParams`, and use the `?` placeholders, which should protect against SQL injection attacks, and then make an issue for switching over to using prepared statements?

Oh okay I misunderstood how the library handled `Exec` and `Query` then haha. Yeah I agree, it'll be a good idea to use the placeholders and the `interpolateParams` parameter for now since it'll make it an easier switch to using prepared statements later down the road. Could you make that issue then?

igorveselinovic committed:
Use backticks to avoid escaping double quotes

igorveselinovic committed:
Update handling of RowsAffected() errors

When it happens after one operation, like in an update, then it is just
logged and not returned.
When it happens in a transaction, like in a create, it is logged and
returned, and the transaction is rolled back.

igorveselinovic committed:
Change helper functions to be methods on Env

igorveselinovic committed:
Update routes to use new format

/{service}/{version}/...

igorveselinovic committed:
Update db queries to use driver's interpolation

This should protect against SQL injection attacks, as opposed to just
building the queries manually which was susceptible to them.
